taskvine/src/manager/vine_manager.c

```diff
@@ -2625,154 +2626,118 @@ struct rmsummary *vine_manager_choose_resources_for_task(struct vine_manager *q,
+		rmsummary_merge_override_basic(limits, t->resources_requested);
+		return limits;
+	}
+
+	/* The total resources we take into account for resource estimate */
+	int64_t cores_total = w->resources->cores.total;
+	int64_t memory_total = w->resources->memory.total;
+	int64_t disk_total = w->resources->disk.total - BYTES_TO_MEGABYTES(w->inuse_cache);
+	int64_t gpus_total = w->resources->gpus.total;
+
+	/* The current available resources on this worker */
+	int64_t cores_available = w->resources->cores.total - w->resources->cores.inuse;
+	int64_t memory_available = w->resources->memory.total - w->resources->memory.inuse;
+	int64_t disk_available = w->resources->disk.total - w->resources->disk.inuse;
+	int64_t gpus_available = w->resources->gpus.total - w->resources->gpus.inuse;
+
+	/* Shortcut if the ramp down mode is enabled. In this case, use all the free space of that worker if the user has not specified.
+	 * Note that we don't use resource_submit_multiplier, as by definition in ramp down there are more workers than tasks. */
+	if (vine_schedule_in_ramp_down(q)) {
+		limits->cores = cores_available;
+		limits->memory = memory_available;
+		limits->disk = disk_available;
+		limits->gpus = gpus_available;
+		rmsummary_merge_override_basic(limits, t->resources_requested);
+		return limits;
 	}
 
+	/* When arrive here, we estimate some of the unspecified resources. Different strategies can be used. */
 	/* Compute the minimum and maximum resources for this task. */
 	const struct rmsummary *min = vine_manager_task_resources_min(q, t);
 	const struct rmsummary *max = vine_manager_task_resources_max(q, t);
 
-	/* available disk for all sandboxes */
-	int64_t available_disk = w->resources->disk.total - BYTES_TO_MEGABYTES(w->inuse_cache);
-
-	/* do not count the size of input files as available.
-	 * TODO: efficiently discount the size of files already at worker. */
-	if (t->input_files_size > 0) {
-		available_disk -= t->input_files_size;
-	}
-
+	/* Use the max as the limits */
 	rmsummary_merge_override_basic(limits, max);
 
-	int use_whole_worker = 1;
-
-	int proportional_whole_tasks = q->proportional_whole_tasks;
-	if (t->resources_requested->memory > -1 || t->resources_requested->disk > -1) {
-		/* if mem or disk are specified explicitely, do not expand resources to fill an integer number of tasks. With this,
-		 * the task is assigned exactly the memory and disk specified. We do not do this for cores and gpus, as the use case
-		 * here is to specify the number of cores and allocated the rest of the resources evenly. */
-		proportional_whole_tasks = 0;
-	}
+	/* Never go below min resources. */
+	rmsummary_merge_max(limits, min);
 
-	/* Proportionally assign the worker's resources to the task if configured. */
 	if (q->proportional_resources) {
-
 		/* Compute the proportion of the worker the task shall have across resource types. */
 		double max_proportion = -1;
-		double min_proportion = -1;
```

Comment:
Min proportion is needed when using automatic resource allocation via categories, please do not remove it.